Implement invariant curl form of contravariant vectors in 3D

erik-f opened the pull request.

sloede commented on src/solvers/dg_curved/containers_3d.jl:
```diff
@@ -68,16 +68,76 @@ function calc_jacobian_matrix!(jacobian_matrix::AbstractArray{<:Any,6}, element,
+  # Jaⁱₙ = 0.5 * ( ∇ × (Xₘ ∇ Xₗ - Xₗ ∇ Xₘ) )ᵢ  where (n, m, l) cyclic and ∇ = (∂/∂ξ, ∂/∂η, ∂/∂ζ)ᵀ
+
+  # Calculate the first summand of the cross product in each dimension
+  for dims in 0:2, j in 1:nnodes(basis), i in 1:nnodes(basis)
```
Maybe you could implement an `eachnode(basis)` and replace the `1:nnodes(basis)` occurrences here and in other parts of the implementaiton?

sloede replied:
Instead of `dims in 0:2` and then later always adding ` +1`, couldn't you just let `dims in 1:3`?

sloede reviewed (changes requested):
In general, this looks ok to me, although I have not verified the algorithms for correctness nor for optimal Julianness. If you want either of those, please request Andrew or Hendrik. Otherwise, this looks OK to me to be merged once the two minor comments are resolved.

sloede commented on src/solvers/dg/basis_lobatto_legendre.jl:
```diff
@@ -74,6 +74,8 @@ end
 
 @inline nnodes(basis::LobattoLegendreBasis{RealT, NNODES}) where {RealT, NNODES} = NNODES
 
+@inline eachnode(basis::LobattoLegendreBasis) = Base.OneTo(nnodes(basis))
```
Maybe you can also replace existing uses of `eachindex(nodes)` with `eachnode(basis)`? I found a few occasions with `grep -r "eachindex.nodes"`.

erik-f replied:
Alright, now there is only one occurrence of `eachindex(nodes)` where `nodes` is passed to a function and `basis` is not defined.

sloede reviewed (changes requested):
One last comment from my side, other than that it LGTM. Maybe you can also tell Andrew and Hendrik what kind of review you're looking for from each of them.

ranocha commented on src/solvers/dg_curved/containers_3d.jl:
```diff
@@ -68,25 +67,83 @@ function calc_jacobian_matrix!(jacobian_matrix::AbstractArray{<:Any,6}, element,
+
+    # Calc only the first summand 0.5 * (Xₘ Xₗ_ζ - Xₗ Xₘ_ζ)_η of
+    # Ja¹ₙ = 0.5 * [ (Xₘ Xₗ_ζ - Xₗ Xₘ_ζ)_η - (Xₘ Xₗ_η - Xₗ Xₘ_η)_ζ ]
+    @views mul!(contravariant_vectors[1, n, i, :, j, element],
+                basis.derivative_matrix,
+                0.5 * (node_coordinates[m, i, :, j, element] .* jacobian_matrix[l, 3, i, :, j, element] .-
+                       node_coordinates[l, i, :, j, element] .* jacobian_matrix[m, 3, i, :, j, element]))
```
You're still allocating for each iteration but that's probably not important here since this function is not performance-critical.

erik-f replied:
I tried to avoid allocations by using `mul!`. I tried to find a non-allocating way to write the multiplication of `node_coordinates` and `jacobian_matrix`, but I didn't find one that doesn't make this less readable than it already is.

In fact, I just did a few benchmarks, and it turns out that the version with `mul!` actually allocates more (and is a little slower) than a simple multiplication.
I decided to use a simple multiplication instead now, since it's more readable and even a bit faster.

sloede replied:
Sounds ok to me... feel free to mark this as resolved

ranocha reviewed (approved):
Looks good to me from a Julia-point-of view. In particular, I like `eachnode(basis)`. I didn't check the correctness of the implementation - that's something for @andrewwinters5000, I guess.
Thanks, @erik-f :+1:

sloede commented:
Here are some comments I received from @andrewwinters5000 this morning:
> I  don’t have time for a full review, but I looked at the code and here are my thoughts:
> 1. They should add a test to very the free-stream preservation
> 2. In principle the indexing and formulas look correct
> 3. It appears that the mapping of the geometry will always use the same polynomial order as the DG approximation. This should be commented somewhere in the code and marked with a TODO to generalize later

> 
> I believe the implementation itself is slightly wrong because they do not cast the product back into the polynomial degree `N` space before they compute each of the derivatives. For example, they currently follow the procedure
> 1. compute the derivatives of the mapping (what is called `jacobian_matrix`)
> 2. compute the product, e.g. `node_coordinates[m, i, :, j, element] .* jacobian_matrix[l, 3, i, :, j, element]`
> 3. then they use the derivative matrix to get the vector component
> 
> However, in the second step the resulting product is now a polynomial of degree 2N-1 such that the differentiation with the derivative matrix is no longer exact and there may be aliasing issues. The procedure above should include an additional step:
> 1. compute the derivatives of the mapping (what is called `jacobian_matrix`)
> 2. compute the product and store it is a temp, e.g. `temp = node_coordinates[m, i, :, j, element] .* jacobian_matrix[l, 3, i, :, j, element]`
> 3. use the Vandermonde matrix to cast this temp variable back into the space of polynomials of degree N
> 4. then  use the derivative matrix to get the vector component
> 
> We should double check this with @gregorgassner because it has been a while since I implemented 3D curved elements but I vaguely remember this interpolation step was important but easy to overlook.

sloede reviewed (approved):
This looks OK from my side, but please address the issues raised by @andrewwinters5000 in https://github.com/trixi-framework/Trixi.jl/pull/532#issuecomment-817515799

gregorgassner commented:
I am not sure, if this step should be done:

3. use the Vandermonde matrix to cast this temp variable back into the space of polynomials of degree N

?

gregorgassner commented:
Just as an information, not sure what is used now: but in FLUXO, we are using the curl invariant form. I would recommend to use the same in Trixi.

sloede commented:
For the free-stream preserving test, please use the same test as described in [arXiv:2012:12040](https://arxiv.org/abs/2012.12040):

Andrés M Rueda-Ramírez, Sebastian Hennemann, Florian J Hindenlang, Andrew R Winters, Gregor Gassner,
**An Entropy Stable Nodal Discontinuous Galerkin Method for the resistive MHD Equations. Part II: Subcell Finite Volume Shock Capturing**,
2020.

![image](https://user-images.githubusercontent.com/3637659/114365076-5f2c8280-9b7a-11eb-809e-32816b712e29.png)

sloede commented:
> However, in the second step the resulting product is now a polynomial of degree 2N-1 such that the differentiation with the derivative matrix is no longer exact and there may be aliasing issues. The procedure above should include an additional step:
> 
>     1. compute the derivatives of the mapping (what is called `jacobian_matrix`)
> 
>     2. compute the product and store it is a temp, e.g. `temp = node_coordinates[m, i, :, j, element] .* jacobian_matrix[l, 3, i, :, j, element]`
> 
>     3. use the Vandermonde matrix to cast this temp variable back into the space of polynomials of degree N
> 
>     4. then  use the derivative matrix to get the vector component

This is also described in

Kopriva, D.A. Metric Identities and the Discontinuous Spectral Element Method on Curvilinear Meshes. J Sci Comput 26, 301 (2006). https://doi.org/10.1007/s10915-005-9070-8

![image](https://user-images.githubusercontent.com/3637659/114366178-7ae45880-9b7b-11eb-8039-4d0241f3fd39.png)

As mentioned [above](https://github.com/trixi-framework/Trixi.jl/pull/532#issuecomment-817562201), interpolation is enough (no projection required).

erik-f commented:
Thank you all for your feedback.

1. To implement tests for free stream preservation, we need #531 to be merged. Otherwise, we can only use rectangular meshes right now.
2. We created tests in a different branch, where we merged this PR and #531 together. We tested a simple mapping `mapping(xi, eta, zeta) = (xi, eta, sin(0.5 * pi * xi) + zeta)` and observed errors around `1e-14`. However, we just implemented the test suggested by @sloede above, and we got errors up to `1e-03`. So it seems like there is really something wrong. We will look into it this afternoon (or tomorrow).
3. Don't we do this interpolation step implicitly? We only evaluate the product at the interpolation nodes, doesn't this give us the interpolation already?
4. @gregorgassner Yes, we implemented the invariant curl form.

sloede commented:
>     1. To implement tests for free stream preservation, we need #531 to be merged. Otherwise, we can only use rectangular meshes right now.

OK, then keep working on #531 such that it can be merged as well. We'll merge this once the tests pass, and then you can debug/make FSP work once all relevant components are there.
